test/framework/src/main/java/org/elasticsearch/test/disruption/LongGCDisruptionTest.java

```diff
@@ -0,0 +1,147 @@
+import java.util.regex.Pattern;
+
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThan;
+
+public class LongGCDisruptionTest extends ESTestCase {
+
+    static class LockedExecutor {
+        ReentrantLock lock = new ReentrantLock();
+
+        public void executeLocked(Runnable r) {
+            lock.lock();
+            try {
+                r.run();
+            } finally {
+                lock.unlock();
+            }
+        }
+    }
+
+    public void testBlockingTimeout() throws Exception {
+        final String nodeName = "test_node";
+        LongGCDisruption disruption = new LongGCDisruption(random(), nodeName) {
+            @Override
+            protected Pattern[] getUnsafeClasses() {
+                return new Pattern[]{
+                    Pattern.compile("LockedExecutor")
+                };
+            }
+
+            @Override
+            protected long getStoppingTimeoutInMillis() {
+                return 100;
+            }
+        };
+        final AtomicBoolean stop = new AtomicBoolean();
+        final CountDownLatch underLock = new CountDownLatch(1);
+        final CountDownLatch pauseUnderLock = new CountDownLatch(1);
+        final LockedExecutor lockedExecutor = new LockedExecutor();
+        final AtomicLong ops = new AtomicLong();
+        try {
+            Thread[] threads = new Thread[10];
+            for (int i = 0; i < 10; i++) {
+                // at least one locked and one none lock thread
+                final boolean lockedExec = (i < 9 && randomBoolean()) || i == 0;
+                threads[i] = new Thread(() -> {
+                    while (stop.get() == false) {
+                        if (lockedExec) {
+                            lockedExecutor.executeLocked(() -> {
+                                try {
+                                    underLock.countDown();
+                                    ops.incrementAndGet();
+                                    pauseUnderLock.await();
+                                } catch (InterruptedException e) {
+
+                                }
+                            });
+                        } else {
+                            ops.incrementAndGet();
+                        }
+                    }
+                });
+                threads[i].setName("[" + nodeName + "][" + i + "]");
+                threads[i].start();
+            }
+            // make sure some threads are under lock
+            underLock.await();
+            RuntimeException e = expectThrows(RuntimeException.class, disruption::startDisrupting);
+            assertThat(e.getMessage(), containsString("stopping node threads took too long"));
+        } finally {
+            stop.set(true);
+            pauseUnderLock.countDown();
+        }
+    }
+
+    public void testNotBlockingUnsafeStackTraces() throws Exception {
```

Comment:
I think this test could use some comments explaining what is happening here, it's quite tricky. :smile: